### ISSUE TITLE
ShufflerIterDataPipe.set_shuffle_settings(True) gets overridden by DataLoader

### ISSUE THREAD

**Issue by NicolasHug** [pytorch/data#295]
Sorry if this isn't the right place for this bug report, please let me know.

I think there's an issue with the shuffling API. Doing

```py
dp = ShufflerIterDataPipe(dp)
dp.set_shuffle_settings(True)
```

won't actually shuffle anything unless `shuffle=True` is also passed to DataLoader.

This is because of

https://github.com/pytorch/pytorch/blob/3399876306772738936596f1711ec1b7f41495c8/torch/utils/data/dataloader.py#L231-L232 
and
https://github.com/pytorch/pytorch/blob/3399876306772738936596f1711ec1b7f41495c8/torch/utils/data/graph_settings.py#L30-L35 
and because of the fact that the default `shuffle` value of DataLoader is `False`.

**Comment by ejguan** [comment]
I think this is intentional design based on Philips's requirement that shuffle would be a no-op by default. And, users can turn it on/off after the pipeline is created.
cc: @pmeier

**Comment by NicolasHug** [comment]
> shuffle would be a no-op by default

Hm, I'm not sure this is even the case right now: the default value of the `default` parameter is `True` https://github.com/pytorch/pytorch/blob/3399876306772738936596f1711ec1b7f41495c8/torch/utils/data/datapipes/iter/combinatorics.py#L85

In torchvision, `hint_shuffling()` sets `default` to False, so that users can then explicitly call `dp.set_shuffle_settings(True)`.

But the problem here is that `dp.set_shuffle_settings(True)` is literally a no-op because it gets overridden by the DataLoader.

**Comment by ejguan** [comment]
> In torchvision, `hint_shuffling()` sets `default` to False, so that users can then explicitly call `dp.set_shuffle_settings(True)`.
> 
> But the problem here is that `dp.set_shuffle_settings(True)` is literally a no-op because it gets overridden by the DataLoader.

Right, it depends on where we want users to set shuffle settings. We use this current implementation because users would use DataLoader to provide shuffle option (via sampler) and we still want to provide the similar experience.

**Comment by pmeier** [comment]
> I think this is intentional design based on Philips's requirement that shuffle would be a no-op by default.

We needed a way to disable shuffling at creation time of the datapipe while later being able to enable it through the data loader.

**Comment by NicolasHug** [comment]
Let's go back to my initial issue, sorry if I'm not being clear enough:

```py

dp = ...
dp = ShufflerIterDataPipe(dp)
dp = ...
dp.set_shuffle_settings(True)

...

dataloader = DataLoader(dp)
```

The dataloader is actually not shuffled at all (for the reasons described in my original comment). This is unexpected because:

- one would expect `dp.set_shuffle_settings(True)` to actually shuffle the dataset
- a non-datapipe iterable dataset woudn't require the `shuffle` parameter to be passed to DataLoader - in fact passing `shuffle` would raise an error.

If we decide that the shuffling of datapipes should only be activated through `DataLoader(shuffle=True)`, then there is no reason to have the `set_shuffle_settings()` at all because it does nothing.

**Comment by ejguan** [comment]
There are basically two different use cases we want to support:
- Iterate over DataPipe
- Iterate using DataLoader
For the first use case, users should be able to use `set_shuffle_settings` to control if they want to do shuffling in the pipeline. For the latter use case, `DataLoader` would become the POC to control it.

Do you have any suggestion on how to improve this UX?

**Comment by NicolasHug** [comment]
The `DataLoader` shouldn't silently override what was set on the datapipe. Perhaps one way would be to change the default of `DataLoader`'s `shuffle` default to `None` instead of False. I believe this was the assumption behind using

```py
if shuffle is not None:
    # override here
```

from https://github.com/pytorch/pytorch/blob/3399876306772738936596f1711ec1b7f41495c8/torch/utils/data/graph_settings.py#L30-L35

**Comment by ejguan** [comment]
I see. That makes sense. This would even help the case of non-DataPipe IterableDataset. Thanks for the suggestion, will make a change in DataLoader.

**Comment by ejguan** [comment]
Closing as it's fixed by https://github.com/pytorch/pytorch/pull/75505
